### PR TITLE
Add a reference to the transpiled JS file in POT file comments

### LINKED ISSUE #1894: [GlobalStep] Some contents on the Payment pages and Checkout page fails to translate to the respective language.
## Bug Description:
Some contents on the Payment pages and Checkout page fails to translate to the respective language

## Test Instructions:
https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-2.5.0#translations-related-testing

## Environment:
Woocommerce Version: WooCommerce 5.3
WooCommerce Payments: v2.5.0

PC:
Windows 10, Mac 10.14.6
Chrome(Version 90.0.4430.93) 
Firefox(Version 88.0)
Safari: v13.1.1

## Steps To Reproduce:
1. Create any test site using JN site.
1. Install and activate all the required plugins 
1. Complete setup wizard using United States address as a store address and without selecting WC Payments on Business Details step.
1. Go to Plugins->Add New->Upload plugin.
1. Install and activate WooCommerce Payments plugin v2.5.0.
1. Go to Payments->Set up to trigger KYC flow.
1. Complete the KYC flow.
1. Go to Settings->General.
1. Change language to one of these 4 languages: German, Italian, Spanish, French.
1. Click “Save Changes”.
1. Go to Dashboard > Updates.
1. Click on “Update Translations” button.
1. Now visit Payments->Deposits/Transactions/Disputes/Settings.
1. Note that some content fails to translate to respective language.
1. As a customer visit to Checkout page.
1. Observe that Some contents on the Payment pages and Checkout page fails to translate.

## Actual Result:
Some contents on the Payment pages and Checkout page fails to translate to the respective language.

## Expected Result:
All the contents on the Payment pages and Checkout page should get translated to the respective language.

## Screenshot:

**Isolating the problem (mark completed items with an [x]):**
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

<details>
```
`
### WordPress Environment ###

WordPress address (URL): https://inquisitive-de.jurassic.ninja
Site address (URL): https://inquisitive-de.jurassic.ninja
WC Version: 5.3.0
REST API Version: ✔ 5.3.0
WC Blocks Version: ✔ 5.1.0
Action Scheduler Version: ✔ 3.1.6
WC Admin Version: ✔ 2.2.6
Log Directory Writable: ✔
WP Version: 5.7.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.46 (Unix) OpenSSL/1.0.2g
PHP Version: 7.4.18
PHP Post Max Size: 1 GB
PHP Time Limit: 30
PHP Max Input Vars: 5000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.7.33-0ubuntu0.16.04.1-log
Max Upload Size: 512 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 5.3.0
WC Database Prefix: wp_
Total Database Size: 8.22MB
Database Data Size: 5.76MB
Database Index Size: 2.46MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.05MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.05MB + Index: 0.03MB + Engine InnoDB
wp_ce4wp_abandoned_checkout: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.05MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailchimp_carts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailchimp_jobs: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_custom_fields: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_dynamic_segment_filters: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_feature_flags: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_forms: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_log: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_mapping_to_external_entities: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletters: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_newsletter_links: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_mailpoet_newsletter_option: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_option_fields: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_posts: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_segment: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_templates: Data: 2.52MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_scheduled_tasks: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_scheduled_task_subscribers: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_segments: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_sending_queues: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_settings: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_statistics_clicks: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_mailpoet_statistics_forms: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_statistics_newsletters: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_statistics_opens: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_mailpoet_statistics_unsubscribes: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_mailpoet_statistics_woocommerce_purchases: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_mailpoet_stats_notifications: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_subscribers: Data: 0.02MB + Index: 0.13MB + Engine InnoDB
wp_mailpoet_subscriber_custom_field: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_subscriber_ips: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_subscriber_segment: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_user_flags: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_ms_snippets: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_options: Data: 1.42MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.19MB + Index: 0.16MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_snippets: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 24
mailpoet_page: 1
page: 7
post: 2
product: 21
product_variation: 7
revision: 1
shop_order: 17
shop_order_refund: 3
shop_subscription: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (12) ###

Query Monitor: by John Blackbourn – 3.7.1
Code Snippets: by Code Snippets Pro – 2.14.1
Companion Plugin: by Osk – 1.18
Creative Mail by Constant Contact: by Constant Contact – 1.3.8
Facebook for WooCommerce: by Facebook – 2.5.0
Google Ads & Marketing by Kliken: by Kliken – 1.0.7
Mailchimp for WooCommerce: by Mailchimp – 2.5.1
MailPoet 3 (New): by MailPoet – 3.60.12
WooCommerce Blocks: by Automattic – 5.1.0
WooCommerce Payments: by Automattic – 2.5.0
WooCommerce Subscriptions: by WooCommerce – 3.0.12
WooCommerce: by Automattic – 5.3.0

### Inactive Plugins (2) ###

Akismet Anti-Spam: by Automattic – 4.1.9
Hello Dolly: by Matt Mullenweg – 1.7.2

### Dropin Plugins (1) ###

db.php: Query Monitor Database Class

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #5 - /shop/
Cart: #6 - /cart/
Checkout: #7 - /checkout/
My account: #8 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 3.6.0 (update to version 3.7.0 is available)
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Subscriptions ###

WCS_DEBUG: ✔ No
Subscriptions Mode: ✔ Live
Subscriptions Live URL: https://inquisitive-de.jurassic.ninja
Subscription Statuses: wc-active: 2
WooCommerce Account Connected: ❌ No
Report Cache Enabled: ✔ Yes
Cache Update Failures: ✔ 0 failure

### Store Setup ###

Country / State: United States (US) — California

### Subscriptions by Payment Gateway ###

WooCommerce Payments: wc-active: 2

### Payment Gateway Support ###

WooCommerce Payments: products
refunds
tokenization
add_payment_method
subscriptions
subscription_cancellation
subscription_suspension
subscription_reactivation
subscription_amount_changes
subscription_date_changes
subscription_payment_method_change
subscription_payment_method_change_customer
subscription_payment_method_change_admin
multiple_subscriptions


### Action Scheduler ###

Complete: 76
Oldest: 2021-05-25 06:56:10 +0000
Newest: 2021-05-25 11:30:08 +0000

Pending: 4
Oldest: 2021-05-26 09:49:41 +0000
Newest: 2021-06-25 11:03:31 +0000


### Status report information ###

Generated at: 2021-05-25 13:11:41 +00:00
`
```
</details>

### PR DESCRIPTION
**Pull request by diegocurbelo**
Fixes #1894

Currently the translations for ts/tsx files are not being exported to the JSON file when generating the language packs in translate.wordpress.com (only translations strings from js files are included).

The JSON files in the language pack contain all strings from the POT file grouped by `.js` filename (using the file reference for each string). This PR fixes the problem by adding additional file references to the generated (webpack transpiled) files for each string.

Example POT file:
```
#: client/additional-methods-setup/methods-selector/add-payment-methods-task.js:391
#: client/connect-account-page/modal/index.js:82
#: client/multi-currency-setup/tasks/store-settings-task/index.js:147
#: client/payment-methods/activation-modal.js:47
msgid "Continue"
msgstr ""
```
becomes:
```
#: dist/index.js:1
#: dist/settings.js:1
#: client/additional-methods-setup/methods-selector/add-payment-methods-task.js:391
#: client/connect-account-page/modal/index.js:82
#: client/multi-currency-setup/tasks/store-settings-task/index.js:147
#: client/payment-methods/activation-modal.js:47
msgid "Continue"
msgstr ""
```

More context in the slack thread: p1638277640394300-slack-CGGCLBN58

#### Changes proposed in this Pull Request

This PR extends the POT combination process to add a comment reference with the generated JS asset file (in the `dist` folder) for each javascript/typescript source file (in the `client` folder).

It also changes the `release.js` task to generate `hidden-source-maps` during production build, they are used to generate the final translations POT file and are not included in the zip package.
An alternative would be to use a Webpack plugin to generate the dependencies file ([like this](https://gist.github.com/diegocurbelo/b1905ab3e1cbd5c5549d9f34eca23098)), but using this plugin does not improve the build time over using hidden-source-maps:
- Develop branch:
`npm run build  113.43s user 7.20s system 274% cpu 43.928 total`
- This branch:
`npm run build  114.46s user 6.95s system 281% cpu 43.070 total`

The times for development builds are not affected, source-maps were already being generated for development builds

#### Testing instructions

1. Checkout `develop`
2. Run `npm run build`
3. Open the generated `languages/woocommerce-payments.pot` file
4. Search for `Risk level` and verify that there is only one reference file in the comments (a tsx file):
```
#: client/transactions/list/index.tsx:132
msgid "Risk level"
msgstr ""
```
5. Checkout this branch
6. Run `npm run build` again
7. Open the generated `languages/woocommerce-payments.pot` file
8. Search for `Risk level` again and verify that this time a reference file to the generated file was included:
```
#: dist/index.js:1
#: client/transactions/list/index.tsx:132
msgid "Risk level"
msgstr ""
```

#### Next steps

With this change the language packs will generate a JSON for each generated file (`dist/index.js`, `dist/settings.js`, etc), so and we can remove [wc-payments-translations-loader.php](https://github.com/Automattic/woocommerce-payments/blob/db27ff47f8d7f74b0d8c1fdef2a1ae57f9a8aab8/includes/class-wc-payments-translations-loader.php), the translations loader would no longer be necessary since from now on the JS file from `wp_register_script()` and the JSON hash would match.
I created #3493 to continue with that without delaying this fix.

### CODE REVIEW

**Comment by vbelolapotkov** [comment]
@diegocurbelo would you like me to check anything specific here? Otherwise I'd ask someone from @Automattic/gamma to take a look.

**Comment by diegocurbelo** [comment]
> @diegocurbelo would you like me to check anything specific here? Otherwise I'd ask someone from @Automattic/gamma to take a look.

Not really, added you and @KarlisJ as reviewers because you were in the original slack thread... I'll ask somebody else 👍🏼

**Comment by kalessil** [comment]
@diegocurbelo : I'm up to review/test it today. Adding myself as a reviewer.

**Comment by KarlisJ** [comment]
The code looks great! It's nice and simple, the "what" is always easy to understand, however, I'd advise carefully adding the "why" in the commit message when this is squashed and merged as it is not clear from the code itself.

However, I think that this should be treated only as a short-term solution. The issue with TS files is WooCommerce wise and thus should be addressed at that level. Right now, any other plugin that introduces TS files with translation will have the same problem and potentially won't know that there's already an example for a solution. 

I'm happy to ship this with the next release to fix the issue for customers ASAP, but we should P2 this problem at WC level.

File: bin/combine-pot-files.php [hunk 1]
```diff
@@ -75,9 +75,90 @@ function woocommerce_admin_parse_pot( $file_name ) {
...
+ *     'dist/index.js',
+ *     'dist/tos.js',
+ *   ]
+ * ]
+ *
+ * @return array Mapping of source js files and the generated files that use them.
+ */
+function load_js_transpiling_source_maps() {
+	$mappings = [];
+	foreach ( glob( "dist/*.js.map" ) as $filename ) {
```
Comment (kalessil) [inline bin/combine-pot-files.php:92]: `glob` has a second parameter, which can disable the filenames sorting (slightly faster execution). If sorting is not a requirement, I'd suggest disabling it.

File: bin/combine-pot-files.php [hunk 2]
```diff
@@ -75,9 +75,90 @@ function woocommerce_admin_parse_pot( $file_name ) {
...
+ * [
+ *   'client/card-readers/settings/file-upload.js' => [
+ *     'dist/index.js',
+ *     'dist/tos.js',
+ *   ]
+ * ]
+ *
+ * @return array Mapping of source js files and the generated files that use them.
+ */
+function load_js_transpiling_source_maps() {
```
Comment (kalessil) [inline bin/combine-pot-files.php:90]: Can we make this stricter towards the types, please: `function load_js_transpiling_source_maps(): array`?

File: bin/combine-pot-files.php [hunk 3]
```diff
@@ -75,9 +75,90 @@ function woocommerce_admin_parse_pot( $file_name ) {
...
+}
+
+/**
+ * For each file reference to a javascript/typescript file (from the client folder) in the comments, it adds file
+ * references to the generated files (from the dist folder) that use that particular javascript/typescript as source.
+ *
+ * @param array $js_mappings Mapping of source js files and the generated files that use them.
+ * @param array $translations POT translations (including references/comments).
+ */
+function add_transpiled_filepath_reference_to_comments( $js_mappings, $translations ) {
```
Comment (kalessil) [inline bin/combine-pot-files.php:127]: Can we make this stricter towards the types, please: `function add_transpiled_filepath_reference_to_comments( array $js_mappings, array $translations ): array`?

**Review by kalessil — APPROVED** [review]
LGTM! Like the elegant solution here!
Left comments with possible improvements.

**Review by kalessil — APPROVED** [review]
Looks fantastic! :shipit:

**Comment by diegocurbelo** [comment]
> The code looks great! It's nice and simple, the "what" is always easy to understand, however, I'd advise carefully adding the "why" in the commit message when this is squashed and merged as it is not clear from the code itself.

Will do.

> However, I think that this should be treated only as a short-term solution. The issue with TS files is WooCommerce wise and thus should be addressed at that level. Right now, any other plugin that introduces TS files with translation will have the same problem and potentially won't know that there's already an example for a solution.

I agree, this should be a problem for any WP plugin that uses TypeScript; I'll start drafting a P2 post so we can work on a standardized way of generating the POT file (for all strings from php, javascript and typescript).

**Pull request merged**